### PR TITLE
Refactor and extend BoundingBox

### PR DESCRIPTION
**Pull request by elliottbiondo**
This PR converts the `BoundingBox` class to `BoundingBox<T>`, templated on the floating point precision of the bounds. In addition, +/- max_double values in bounding boxes read from disk are converted into +/- infinities. It also adds new utility classes.

Split from #849

### CODE REVIEW

File: src/orange/BoundingBox.hh
Comment (sethrj) [inline src/orange/BoundingBox.hh]: Hmm, do we use `Real3` or allow it to be more generic? Looking at the class again, there are a couple things that won't compile if we use integers (the nan and infinity accessors). So maybe we should just assume/require it's for real numbers for the time being (sorry to flip-flop on this... I didn't recall the full picture last time).

Even if we leave this as generic, since `array_type` isn't a standard type definition like `value_type` is, maybewe should write as `Array3`?

File: src/orange/BoundingBoxIO.json.hh
Comment (sethrj) [inline src/orange/BoundingBoxIO.json.hh]: Template deduction would be automatic if you wrote
```suggestion
template<class T>
inline void fix_inf(Array<T, 3>* point)
```

File: src/orange/BoundingBoxUtils.hh
Comment (sethrj) [inline src/orange/BoundingBoxUtils.hh]:
```suggestion
inline Array<T, 3> center(BoundingBox<T> const& bbox)
```

File: src/orange/BoundingBoxUtils.hh
Comment (sethrj) [inline src/orange/BoundingBoxUtils.hh]:
```suggestion
    Array<T, 3> center;
```

File: src/orange/BoundingBoxUtils.hh
Comment (sethrj) [inline src/orange/BoundingBoxUtils.hh]:
```suggestion
    Array<T, 3> lengths;
```

File: src/orange/BoundingBoxUtils.hh
Comment (sethrj) [inline src/orange/BoundingBoxUtils.hh]:
```suggestion
    Array<T, 3> lower; 
    Array<T, 3> upper;
```

File: src/orange/BoundingBox.hh
Comment (elliottbiondo) [reply src/orange/BoundingBox.hh]: I'll change it to Array3

File: src/orange/BoundingBoxUtils.hh
Comment (elliottbiondo) [reply src/orange/BoundingBoxUtils.hh]: This one was just an oversight my part

Comment (sethrj) [reply src/orange/BoundingBoxUtils.hh]: That's what reviews are for 😄

**Comment by elliottbiondo** [comment]
@sethrj CI passes, this should be good to go

**Review by sethrj — APPROVED** [review]
Thanks for pulling this changeset out. I'm going to wait until EOBish tomorrow to merge: I want to get the non-API-changing stuff in first.

File: src/orange/BoundingBoxIO.json.hh
Comment (sethrj) [inline src/orange/BoundingBoxIO.json.hh:21]: It's (always?) a good idea to put anonymous namespaces *inside* of the main namespace you're using: that way it can use the same scope as a regular function, but it simply won't generate a symbol.

Comment (elliottbiondo) [reply src/orange/BoundingBoxIO.json.hh:21]: How about I fix it in the next MR so we don't have to rerun CI?

Comment (sethrj) [reply src/orange/BoundingBoxIO.json.hh:21]: It's fixed now. Just an informational message.

**Pull request merged**
